benchmarks/discovery.md

```diff
@@ -0,0 +1,148 @@
+# Discovery Protocol Benchmarks
+
+The goal of this benchmark design is to create a controlled environment in which
+we can gather data on performance, efficiency, and reliability of the two
+network discovery protocols proposed for Open Screen:
+
+* [SSDP](../ssdp.md)
+* [mDNS](../mdns.md)
+
+# Observables
+
+* Network efficiency & utilization
+  * Number of packets sent and received per device per minute
+  * Number of bytes sent and received per device per minute
+  * Total number of packets and bytes sent per minute
+  * Peak number of packets and bytes sent per second
+* Power efficiency
+  * Total mW consumed per minute per device
+* Reliability
+  * Percentage of devices found after 10s, 30s, 60s
+* Latency
+  * Time to find all devices
+  * Time to discover that a device has been connected
+  * Time to discover that a device has been disconnected
+  
+# Test Environment
+
+## Hardware and Operating System
+
+The test environment will consist of 10
+[Raspberry Pi 2 Model B](https://www.adafruit.com/product/2358)
+single-board computers with:
+* ARM Cortex-A53 1.2 GHz quad-core processor
+* 1 GB RAM
+* 802.11 b/g/n wireless LAN module (via USB)
+
+These are similar to Android One smartphones and slightly more powerful than
+previous generation streaming devices (see [Device Specs](../device_specs.md)).
+
+The default operating system for Raspberry Pi devices is
+[Raspbian](https://www.raspberrypi.org/downloads/raspbian/), which is
+Linux-based.  The "Lite" version should be sufficient as we don't need an
+interactive OS.
+
+## Networking
+
+The test devices will be connected via a 802.11/n, 2.4 GHz WiFi network secured
+by WPA2-PSK and provided by an off the shelf, basic home Wi-Fi router.
+
+Typical models:
+* [Netgear N300](https://www.netgear.com/home/products/networking/wifi-routers/WNR3500L.aspx)
+* [TP-Link TL-WR841N](http://www.tp-link.com/us/products/details/cat-5506_TL-WR841N.html)
+* [ASUS RT-N12 D1](https://www.asus.com/us/Networking/RTN12_D1/)
+* [Zyxel NBG-418N v2](http://www.zyxel.com/us/en/products_services/Wireless-N300-Home-Router-NBG-418N-v2/)
+
+For the sake of reproducibility, we will re-run benchmarks with multiple router
+models to see if there is any signficant variation in results.
+
+## Software
+
+For mDNS, we will use the [Avahi](https://www.avahi.org/) open source mDNS
+software that is available as a Debian package.
+
+For SSDP, we will either:
+
+1. Fork and customize
+[libupnp](http://pupnp.sourceforge.net/) for the [Open Screen SSDP discovery
+protocol](../ssdp.md).
+2. Write our own SSDP implementation from scratch, possibly based on the
```

Comment:
Feel free to use this simple Node.js implementation https://github.com/fraunhoferfokus/peer-ssdp/blob/master/lib/peer-ssdp.js in the Benchmark responder test app which works also on Raspberry Pi. It implements only the SSDP protocol and not the whole UPnP Stack. For Android we can use this https://github.com/fraunhoferfokus/cordova-plugin-hbbtv/tree/master/src/android/ssdp class and use it in the Benchmark sender app.

Reply:
Done, added the Node.js as an alternative implementation for the SSDP stack here.  I also added a link to the Android plugin to the section at the end on mobile device benchmarking.